Studying Warburg Effect in cancer cells through visualization of Elementary Flux Modes (EFMs) from Recon 2.2
This code is part of the study:
"An integrative workflow to visualize Elementary Flux Modes in genome-scale metabolic models", Sarathy et al. (2019) (in preparation)

It was used for visualizing the EFMs that were selected after running the script <>. 
First, the the SBML file(s) created in the mlx script are loaded into Cytoscape. 
After which a set of functions from RCy3 are called to perform network layout operations and mapping gene expression on the visualized network. 
This automation allows users to visualize multiple EFMs in shorter time. 
---
title: "efmviz_rcy3_human"
author: "Chaitra Sarathy"
date: "July 11, 2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.

DEPENDENCIES
- R libraries: RCy3
- Cytpscape (v 3.5 and above)
- Cytpscape apps: cy3sbml, yFiles Layout

INPUTS
- EFM saved as an SBML file (as generated from <>)
- Data file: Comma separated text file containing two columns, reaction names (column name - rxnID) and flux values (column name - fluxVector), as generated from <>
- Network style file provided under core/R/<>

OUTPUTS
EFM from E. coli that was selected from <> is visualized as a network of genes, reactions and metabolites with reaction fluxes mapped on the edges in the network.

-----------------------------------------------------------------------------------------------------------------------------------------------------------------------
DO THIS IN THE CYTOSCAPE GUI:

MANUAL STEP 1. Importing SBML file in Cytoscape through cy3sbml
- Open Cytoscape (v 3.5 and above) and make sure cy3sbml app has been installed. If installing for the first time, a Cytoscape restart maybe necessary
- Click on the cy3sbl app-specific icon and select the EFM (SBML submodel file which was generated from MATLAB) to open
Three networks are generated for every model imported and they are respectively identified using the prefixes: All, Base and Kinetic. 
- Click on the Base network and run the script below. (Other two networks, All and Kinetic may be deleted from GUI)
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------
Before running this script, declare file locations and file names. 
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------
NOTE: CHANGE THE NAMES OF ONLY THESE VARIABLES
```{r}
usecase2_fileName = "C:/Users/chaitra.sarathy/Dropbox/Chaitra/Analysis/EFM/visualizationOfEFMs/githubVersion/data/data_useCase_2.xlsx"
styleName = "efmviz_style"
```
Install and load necessary R packages
```{r}
if("RCy3" %in% rownames(installed.packages()) == FALSE) {install.packages("RCy3")}
if("readxl" %in% rownames(installed.packages()) == FALSE) {install.packages("readxl")}
if("RColorBrewer" %in% rownames(installed.packages()) == FALSE) {install.packages("RColorBrewer")}
if("rstudioapi" %in% rownames(installed.packages()) == FALSE) {install.packages("rstudioapi")}

library(RCy3)
library(readxl)
library(RColorBrewer)
library(rstudioapi)
```

Make sure the connection between R and Cytoscape works
```{r}
cytoscapePing() #must return "You are connected to Cytoscape!"
```
cytoscapePing() must return "You are connected to Cytoscape!". If not, Cytoscape needs a restart.

MAKE SURE THE BASE NETWORK IS SELECTED IN CYTOSCAPE

Import and apply the provided style file 
```{r}
setVisualStyle(style.name = styleName)
```

Updates to the style that was applied above
# Color reaction nodes as green
```{r}
clearSelection()
setNodeColorBypass(node.names = getSelectedNodes(selectNodes(nodes = "reaction", by.col = "sbml type")), new.colors = "#00FF00" )
``` 

Get node table of the network to modify node display labels - add gene symbols and remove lengthy reaction IDs which are then mapped back on the network

# Get three columns label (default value of label), name (node identifier) and sbml type (for filtering nodes later)
```{r}
nodeTab = getTableColumns(table = "node", columns = c("label", "sbml type", "name"))
nodeTab = data.frame(nodeTab, stringsAsFactors = F)
```

# Import gene symbol information
```{r}
human_geneNames = read_excel(usecase2_fileName, sheet = 2)
```

# Add gene symbols
```{r}
nodeTab[,"nodeLabels"] = character()
nodeTab$nodeLabels = nodeTab$label
nodeTab$nodeLabels[which(nodeTab$label %in% human_geneNames$HGNCID)] = human_geneNames$GeneSymbol[which(human_geneNames$HGNCID %in% nodeTab$label)]
```

# Remove reaction label
```{r}
nodeTab$nodeLabels[which(nodeTab$`sbml type` == "reaction")] = ""
```
# Load the updated node table into Cytoscape 
```{r}
loadTableData(data = nodeTab, data.key.column = "label", table = "node", table.key.column = "label")
```
# Remap node labels to new label
```{r}
updateStyleMapping(style.name = styleName, mapVisualProperty('node label', 'nodeLabels','p'))
```

Add gene expression data
# Import the file containing gene expresssion data
```{r}
cancerGEx = read_excel(usecase2_fileName, sheet = 1)
cancerGEx[,"HGNCID"] = character()
cancerGEx$HGNCID[which(cancerGEx$GeneID %in% human_geneNames$EnsemblID)] = human_geneNames$HGNCID[which(human_geneNames$EnsemblID %in% cancerGEx$GeneID)]
```

# merge both tables
```{r}
nodeTab_withData = merge(x=nodeTab, y=cancerGEx, by.x = "label", by.y="HGNCID", all.x=T)
nodeTab_withData = nodeTab_withData[!is.na(nodeTab_withData$log2FC),]
```
# Export table to cytoscape
```{r}
loadTableData(data = nodeTab_withData, data.key.column = "label", table = "node", table.key.column = "label")
```

Map gene expression to the gene nodes
```{r}
setNodeColorMapping('log2FC', c(-2,0,2), c("#3182bd","#f5f5f5","#de2d26"), style.name=styleName)
```

# show gene pvalue on edge weight
# createColumnFilter(filter.name = "geneEx_sigPVal", column = "P.Value", criterion = 0.05, predicate = "LESS_THAN", caseSensitive = FALSE)
# setNodeBorderWidthBypass(node.names = getSelectedNodes(), new.widths = 8)
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------
DO THIS IN THE CYTOSCAPE GUI:
MANUAL STEP 2. To apply layout, click Layout --> yFiles Orthogonal Layout